runtime/domain.c

```diff
@@ -1363,8 +1365,16 @@ int caml_domain_is_in_stw(void) {
       finishes.
    The same logic would apply for any other situations in which a domain
    wants to join or leave the set of STW participants.
+
+  The explanation above applies if [sync] = 1. When [sync] = 0, no
+  synchronization happens, and we simply run the handler asynchronously on
+  all domains. We still hold the stw_leader field until we know that
+  every domain has run the handler, so another STW section cannot
+  interfere with this one.
+
 */
 int caml_try_run_on_all_domains_with_spin_work(
+  int sync,
```

Comment:
I'm not entirely sure we want to change this interface for this one use case. This function looks deceptively simple to use but actually ended up being a source of many bugs.

I think there is one in the use here: https://github.com/ocaml/ocaml/pull/11903/files#diff-67115925103982a8ebeb085cfab5ef31a182c9a442bc51e053934364d3750dafR1637 . `caml_try_run_on_all_domains_with_spin_work` can just fail to actually run the function if it can't be the stw_leader or take the all_domains_lock. This means that we might have a `requested_global_major_slice` whose actions actually get ignored.

As an alternative we could support the use-case here with something simpler:

```
while(1) {
  handle_interrupts();
  if( caml_plat_try_lock(&all_domains_lock) ) {
   /* iterate through stw_domains.participating_domains and set requested_major_slice to 1 */
   caml_plat_unlock(&all_domains_lock);
  }
}
```

The only race is when you're about to enter a major slice anyway, so doesn't make a difference.